### PR TITLE
Integrate NVTabular into Morpheus Core and replace existing column_info based workflows.

### LINKED ISSUE #862: NVTabular Updates to Morpheus Core
We should update the `morpheus.utils.column_info` library to use NVTabular since the functionality is nearly identical but NVT will provide more features that are regularly tested

```[tasklist]
- [x] Update dependencies to include `nvtabular` in the conda environement
- [ ] https://github.com/nv-morpheus/Morpheus/issues/863
- [x] Update documentation around new changes to `morpheus.utils.column_info`
```

### LINKED ISSUE #863: Replace the classes and functions in `morpheus.utils.column_info` with equivalents from NVTabular
1. Map all implementations of `ColumnInfo` to NVTabular operation equivalents
      1. If any cannot be mapped, create custom operations
   2. Replace all uses of the `DataFrameInputSchema` class with `nvt.Schema`
      1. This could also use `nvt.Workflow` potentially
   3. Replace all uses of `process_dataframe` with `nvt.Workflow.fit_transform()`
   4. Add metadata tags for use in specific pipelines implementations (i.e. UserID column tags, date tags, etc.)
   5. Ideally, we would keep this as backwards compatible as possible keeping the existing class names and public API and replace the implementation with NVT

### PR DESCRIPTION
**Pull request by drobison00**
## Description

Closes #862 
Closes #863 

Integrates NVT into Morpheus core
- Appropriate conda packages added
- Existing process_dataframe calls in column_info.py are converted to nvt.Workflows
- Backwards compatibility is maintained for anyone using DataFrameInputSchema objects

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/nv-morpheus/Morpheus/blob/main/docs/source/developer_guide/contributing.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: morpheus/utils/schema_transforms.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    if (input_schema.row_filter is None):
+        return df_in
+
+    return input_schema.row_filter(df_in)
+
+
+def process_dataframe(
+    df_in: typing.Union[pd.DataFrame, cudf.DataFrame],
+    input_schema: typing.Union[nvt.Workflow, DataFrameInputSchema],
+) -> pd.DataFrame:
```
Comment (mdemoret-nv) [inline morpheus/utils/schema_transforms.py:122]: Add overload information for this function:

```suggestion
@typing.overload
def process_dataframe(
    df_in: pd.DataFrame,
    input_schema: typing.Union[nvt.Workflow, DataFrameInputSchema],
) -> pd.DataFrame:
    ...
@typing.overload
def process_dataframe(
    df_in: cudf.DataFrame,
    input_schema: typing.Union[nvt.Workflow, DataFrameInputSchema],
) -> cudf.DataFrame:
    ...
def process_dataframe(
    df_in: typing.Union[pd.DataFrame, cudf.DataFrame],
    input_schema: typing.Union[nvt.Workflow, DataFrameInputSchema],
):
```

File: morpheus/utils/nvt/schema_converters.py [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+    :return: A MutateOp object that represents the JSON flatten operation
+    """
+    json_flatten_op = MutateOp(json_flatten, dependencies=json_input_cols, output_columns=json_output_cols)
+
+    return json_flatten_op
+
+
+@sync_df_as_pandas
+def string_cat_col(df: typing.Union[pd.DataFrame, cudf.DataFrame], output_column,
+                   sep) -> typing.Union[pd.DataFrame, cudf.DataFrame]:
```
Comment (mdemoret-nv) [inline morpheus/utils/nvt/schema_converters.py:143]: With the decorator, this will only ever take pandas objects

File: morpheus/utils/nvt/schema_converters.py [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+    :param input_columns: The input columns to concatenate
+    :param sep: The separator to use when concatenating the strings
+    :return: The resulting DataFrame
+    """
+    return string_cat_col(df[input_columns], output_column=output_column, sep=sep)
+
+
+@sync_df_as_pandas
+def increment_column(df: typing.Union[pd.DataFrame, cudf.DataFrame], output_column, input_column, period: str = 'D') \
+        -> typing.Union[pd.DataFrame, cudf.DataFrame]:
```
Comment (mdemoret-nv) [inline morpheus/utils/nvt/schema_converters.py:177]: With the decorator, this will only ever take pandas objects

File: morpheus/utils/nvt/schema_converters.py [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+# Mappings from ColumnInfo types to functions that create the corresponding NVT operator
+ColumnInfoProcessingMap = {
+    BoolColumn:
+        lambda ci,
+        deps: [
+            LambdaOp(
+                lambda series: series.map(ci.value_map).astype(bool), dtype="bool", label=f"[BoolColumn] '{ci.name}'")
+        ],
+    # ColumnInfo: lambda ci, deps: [
+    #    LambdaOp(lambda series: series.astype(ci.dtype), dtype=ci.dtype, label=f"[ColumnInfo] '{ci.name}'")],
```
Comment (mdemoret-nv) [inline morpheus/utils/nvt/schema_converters.py:210]: Excess comment

File: morpheus/utils/nvt/schema_converters.py [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+                     output_columns=[(ci.name, ci.dtype)],
+                     label=f"[StringJoinColumn] '{ci.input_name}' => '{ci.name}'")
+        ],
+    JSONFlattenInfo:
+        lambda ci,
+        deps: [json_flatten_from_input_schema(ci.input_col_names, ci.output_col_names)]
+}
+
+
+def build_nx_dependency_graph(column_info_objects: typing.List[ColumnInfo]) -> nx.DiGraph:
```
Comment (mdemoret-nv) [inline morpheus/utils/nvt/schema_converters.py:280]: Need a docstring

Comment (mdemoret-nv) [inline morpheus/utils/nvt/schema_converters.py:280]: If this is only used in this file and in tests, prefix with an underscore to show users shouldnt use this function.

File: morpheus/utils/nvt/schema_converters.py [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+            for input_col_name in col_info.input_columns:
+                dep_col_info = find_dependent_column(input_col_name, col_info.name)
+                if dep_col_info:
+                    graph.add_edge(dep_col_info.name, col_info.name)
+
+        elif col_info.__class__ == JSONFlattenInfo:
+            for output_col_name in [c for c, _ in col_info.output_col_names]:
+                dep_col_info = find_dependent_column(output_col_name, col_info.name)
+                if dep_col_info:
+                    graph.add_edge(dep_col_info.name, col_info.name)
```
Comment (mdemoret-nv) [inline morpheus/utils/nvt/schema_converters.py:315]: Can this functionality be baked into the ColumnInfo classes themselves? Like a virtual function that returns the dependent column info? This would help locate the functionality with the class

File: morpheus/utils/nvt/schema_converters.py [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+
+    return graph
+
+
+def bfs_traversal_with_op_map(graph, ci_map, root_nodes):
```
Comment (mdemoret-nv) [inline morpheus/utils/nvt/schema_converters.py:320]: What are the data types for these arguments? Also needs a docstring

File: morpheus/utils/nvt/schema_converters.py [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+
+            # Add our neighbors to the queue
+            neighbors = [n for n in graph.neighbors(node)]
+            for neighbor in neighbors:
+                queue.append(neighbor)
+
+    return visited, node_op_map
+
+
+def coalesce_leaf_nodes(node_op_map, graph, preserve_re):
```
Comment (mdemoret-nv) [inline morpheus/utils/nvt/schema_converters.py:363]: What are the data types for these arguments? Also needs a docstring

File: morpheus/utils/nvt/schema_converters.py [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+        if len(neighbors) == 0 or (preserve_re and preserve_re.match(node)):
+            if coalesced_workflow is None:
+                coalesced_workflow = op
+            else:
+                coalesced_workflow = coalesced_workflow + op
+
+    return coalesced_workflow
+
+
+def coalesce_ops(graph, ci_map, preserve_re=None):
```
Comment (mdemoret-nv) [inline morpheus/utils/nvt/schema_converters.py:377]: What are the data types for these arguments? Also needs a docstring

File: morpheus/utils/nvt/schema_converters.py [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+    # Try to guess which output columns we'll produce
+    json_output_cols = resolve_json_output_columns(input_schema)
+
+    json_cols = input_schema.json_columns
+    column_info_objects = [ci for ci in input_schema.column_info]
+    if (json_cols is not None and len(json_cols) > 0):
+        column_info_objects.append(
+            JSONFlattenInfo(
+                input_col_names=[c for c in json_cols],
+                # output_col_names=[name for name, _ in json_output_cols],
```
Comment (mdemoret-nv) [inline morpheus/utils/nvt/schema_converters.py:408]: Extra comment

File: morpheus/utils/nvt/schema_converters.py [hunk 11]
```diff
@@ -0,0 +1,430 @@
...
+
+    column_info_map = {ci.name: ci for ci in column_info_objects}
+
+    graph = build_nx_dependency_graph(column_info_objects)
+
+    # Uncomment to print the computed: dependency layout
+    # from matplotlib import pyplot as plt
+    # pos = graphviz_layout(graph, prog='neato')
+    # nx.draw(graph, pos, with_labels=True, font_weight='bold')
+    # plt.show()
```
Comment (mdemoret-nv) [inline morpheus/utils/nvt/schema_converters.py:421]: Maybe switch this to work off an environment variable to print the debug info when needed?

File: morpheus/utils/nvt/transforms.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+# limitations under the License.
+import json
+import typing
+
+import pandas as pd
+
+import cudf
+
+
+def json_flatten(col_selector, df: typing.Union[pd.DataFrame, cudf.DataFrame]):
```
Comment (mdemoret-nv) [inline morpheus/utils/nvt/transforms.py:22]: Can this use the `sync_df_as_pandas` decorator?

File: tests/test_column_info.py [hunk 13]
```diff
@@ -65,42 +66,47 @@ def test_dataframe_input_schema_with_json_cols():
...
         StringCatColumn(name="location",
-                        dtype=str,
+                        dtype='str',
                         input_columns=[
                             "properties.location.city",
                             "properties.location.countryOrRegion",
                         ],
                         sep=", "),
-        RenameColumn(name="statusfailureReason", dtype=str, input_name="properties.status.failureReason"),
+        RenameColumn(name="statusfailureReason", dtype='str', input_name="properties.status.failureReason"),
```
Comment (mdemoret-nv) [inline tests/test_column_info.py:86]: I think these should use their original type object to test conversion to strings

File: tests/test_column_info.py [hunk 14]
```diff
@@ -65,42 +66,47 @@ def test_dataframe_input_schema_with_json_cols():
...
     assert "userId" in processed_df_cols
     assert "time" not in processed_df_cols
     assert "properties.userPrincipalName" not in processed_df_cols
 
+    # Test that we get the same answer when the dataframe is processed via workflow
+    # TODO: Uncomment once https://github.com/rapidsai/cudf/issues/13305 is fixed
+    # nvt_workflow = dataframe_input_schema_to_nvt_workflow(schema)
+    # df_processed_workflow = process_dataframe(input_df, nvt_workflow)
+    # assert df_processed_schema.equals(df_processed_workflow)
+
```
Comment (mdemoret-nv) [inline tests/test_column_info.py:106]: Should this be uncommented?

File: tests/test_column_info.py [hunk 15]
```diff
@@ -162,8 +168,58 @@ def test_dataframe_input_schema_without_json_cols():
 
 
 @pytest.mark.use_python
-def test_string_cat_column():
+def test_dataframe_input_schema_without_json_cols_nvt():
+    src_file = os.path.join(TEST_DIRS.tests_data_dir, "azure_ad_logs.json")
```
Comment (mdemoret-nv) [inline tests/test_column_info.py:172]: Use the `DatasetManager` class to load dataframes

File: tests/utils/nvt/integration/test_mutate_op.py [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+        '{"key1": "value3", "key2": {"subkey1": "subvalue5", "subkey2": "subvalue6"}}'
+    ]
+
+    expected_pdf = pd.DataFrame({
+        'col1.key1': ['value1', 'value2', 'value3'],
+        'col1.key2.subkey1': ['subvalue1', 'subvalue3', 'subvalue5'],
+        'col1.key2.subkey2': ['subvalue2', 'subvalue4', 'subvalue6']
+    })
+
+    return json_data, expected_pdf
```
Comment (mdemoret-nv) [inline tests/utils/nvt/integration/test_mutate_op.py:37]: Make this a fixture

File: tests/utils/nvt/test_json_flatten_transform.py [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+from morpheus.utils.nvt.transforms import json_flatten
+
+
+def test_json_flatten_pandas():
+    data = {
+        "id": [1, 2],
+        "info": [
+            '{"name": "John", "age": 30, "city": "New York"}', '{"name": "Jane", "age": 28, "city": "San Francisco"}'
+        ]
+    }
```
Comment (mdemoret-nv) [inline tests/utils/nvt/test_json_flatten_transform.py:29]: Make this data a fixture since its used more than once

File: tests/utils/nvt/test_mutate_op.py [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+def setUp():
+    df = pd.DataFrame({'A': [1, 2, 3], 'B': [4, 5, 6], 'C': [7, 8, 9]})
+
+    def example_transform(col_selector: ColumnSelector, df: DataFrameType) -> DataFrameType:
+        selected_columns = col_selector.names
+        for col in selected_columns:
+            df[col + '_new'] = df[col] * 2
+        return df
+
+    return df, example_transform
```
Comment (mdemoret-nv) [inline tests/utils/nvt/test_mutate_op.py:34]: Make this a fixture

File: tests/utils/nvt/test_transforms.py [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+from morpheus.utils.nvt.transforms import json_flatten
+
+
+def test_json_flatten_pandas():
+    data = {
+        "id": [1, 2],
+        "info": [
+            '{"name": "John", "age": 30, "city": "New York"}', '{"name": "Jane", "age": 28, "city": "San Francisco"}'
+        ]
+    }
```
Comment (mdemoret-nv) [inline tests/utils/nvt/test_transforms.py:29]: Make a fixture for this.

File: morpheus/utils/nvt/schema_converters.py [hunk 20]
```diff
@@ -0,0 +1,430 @@
...
+            df = df.to_pandas()
+
+        df = func(df, **kwargs)
+
+        if convert_to_cudf:
+            df = cudf.from_pandas(df)
+
+        return df
+
+    return wrapper
```
Comment (mdemoret-nv) [inline morpheus/utils/nvt/schema_converters.py:63]: This decorator can be improved using some new Python 3.10 typing functionality:
```suggestion
XDataFrame = typing.TypeVar("XDataFrame", pd.DataFrame, cudf.DataFrame)
"""
Represents a DataFrame that can be either a pandas or cudf DataFrame
"""

_SyncPandasArgs = typing.ParamSpec('_SyncPandasArgs')
"""
Represents the remaining arguments to the function after the first argument (the DataFrame)
"""


def sync_df_as_pandas(
    func: typing.Callable[typing.Concatenate[pd.DataFrame, _SyncPandasArgs], pd.DataFrame]
) -> typing.Callable[typing.Concatenate[XDataFrame, _SyncPandasArgs], XDataFrame]:
    """
    This function serves as a decorator that synchronizes cudf.DataFrame to pandas.DataFrame before applying the function.

    :param func: The function to apply to the DataFrame
    :return: The wrapped function
    """

    def wrapper(df: XDataFrame, *args: _SyncPandasArgs.args, **kwargs: _SyncPandasArgs.kwargs) -> XDataFrame:
        convert_to_cudf = False
        if type(df) == cudf.DataFrame:
            convert_to_cudf = True
            df = df.to_pandas()

        df = func(df, *args, **kwargs)

        if convert_to_cudf:
            df = cudf.from_pandas(df)

        return df

    return wrapper

```

**Review by mdemoret-nv — CHANGES_REQUESTED** [review]
General comments:

1. Wrong format for docstrings. Make sure to use NumpyDoc format: https://numpydoc.readthedocs.io/en/latest/example.html
2. Would like to see some tests added to exercise the type -> str conversion on `ColumnInfo` to make sure there are no edge cases.

File: morpheus/utils/schema_transforms.py [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+
+from morpheus.utils.column_info import DataFrameInputSchema
+from morpheus.utils.nvt import dataframe_input_schema_to_nvt_workflow
+# Apply patches to NVT
+# TODO(Devin): Can be removed, once numpy mappings are updated in Merlin
+# ========================================================================
+from morpheus.utils.nvt.patches import patch_numpy_dtype_registry
+
+patch_numpy_dtype_registry()
+# ========================================================================
```
[the remark this answers is not in the document]
Comment (drobison00) [reply morpheus/utils/schema_transforms.py:31]: Yes, this isn't related to any bugs, its just adding a conversion path that isn't in the upstream. If you don't have it some conversions will fail in NVT.

File: morpheus/utils/nvt/__init__.py [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+    from merlin.dtypes.mapping import DTypeMapping
+
+    morpheus_extension = DTypeMapping(mapping={
+        mn.datetime64: [datetime],
+    }, )
+
+    register("morpheus_ext", morpheus_extension)
+
+
+register_morpheus_extensions()
```
[the remark this answers is not in the document]
Comment (drobison00) [reply morpheus/utils/nvt/__init__.py:33]: Its adding a custom converter, that lets merlin datetime64 dtypes know they can be converted to python datetime objects.

File: morpheus/utils/schema_transforms.py [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+        return cudf.from_pandas(df_normalized)
+
+    return df_normalized
+
+
+def _filter_rows(df_in: pd.DataFrame, input_schema: DataFrameInputSchema):
+    if (input_schema.row_filter is None):
+        return df_in
+
+    return input_schema.row_filter(df_in)
```
[the remark this answers is not in the document]
Comment (drobison00) [reply morpheus/utils/schema_transforms.py:116]: No, you're right, was keeping them around for comparative testing. Now that all the unit tests are passing I'll remove them.

File: morpheus/utils/nvt/schema_converters.py (hunk 6 above)
Comment (drobison00) [reply morpheus/utils/nvt/schema_converters.py:315]: I'll take a look at this; I'd originally tried to avoid modifying the ColumnInfo classes, but based on our discussions, it sounds like we might want to continue using them going forward and just translate to NVT.

File: morpheus/utils/nvt/transforms.py (hunk 12 above)
Comment (drobison00) [reply morpheus/utils/nvt/transforms.py:26]: I've tinkered with this a bit and ran into some problems with combining it with functools.partial, because of arg/kwarg orderings, and I can't change the signature because its used by NVT as an operator. I'll come back to this if there is time.

File: tests/test_column_info.py (hunk 13 above)
Comment (drobison00) [reply tests/test_column_info.py:86]: That's not really their dtype though, its something that gets translated into a pandas/cudf type. It also gets returned by the ColumnInfo 'get_pandas_dtype' call.

File: morpheus/utils/nvt/patches/merlin_patches.py [hunk 24]
```diff
@@ -0,0 +1,33 @@
...
+
+    Until this is fixed upstream, with the mappings added to `merlin/dtypes/mappings/numpy.py`, this patch should be
+    used. The function is idempotent, and should be called before any NVT operators are used.
+    :return:
+    """
+    from merlin.dtypes import _dtype_registry
+
+    numpy_dtypes = _dtype_registry.mappings["numpy"].from_merlin_
+    if (mn.struct not in numpy_dtypes.keys()):
+        numpy_dtypes[mn.struct] = [np.dtype("O"), object]
```
[the remark this answers is not in the document]
Comment (drobison00) [reply morpheus/utils/nvt/patches/merlin_patches.py:33]: I force it to be called in the area where its used, because its not universally required. Also, since its kind of a hacky patch, and requires us to import some merlin libraries, I didn't want to hide it in the init file and silently call it.

If anyone else writes NVT operators that need to make cudf struct -> numpy object conversions, they'll encounter a failure condition.

File: morpheus/utils/column_info.py [hunk 25]
```diff
@@ -69,10 +75,14 @@ class ColumnInfo:
...
-        if (issubclass(self.dtype, datetime)):
+        if ((isinstance(self.dtype, str) and self.dtype.startswith("datetime"))
+                or (isinstance(self.dtype, type) and issubclass(self.dtype, datetime))):
             return "datetime64[ns]"
         else:
-            return self.dtype
+            if (isinstance(self.dtype, str)):
+                return self.dtype
+            else:
+                return self.dtype.__name__
```
[the remark this answers is not in the document]
Comment (drobison00) [reply morpheus/utils/column_info.py]: get_pandas_dtype is called in a number of places, and has a different API. I'll look at adding this in a future pass.

File: morpheus/utils/nvt/schema_converters.py (hunk 6 above)
Comment (drobison00) [reply morpheus/utils/nvt/schema_converters.py]: I want to spend more time thinking about the best way to do this. If we're going to use ColumnInfo to define workflows, lets save this for the next round of updates.

File: docker/conda/environments/cuda11.8_dev.yml [hunk 26]
```diff
@@ -15,11 +15,12 @@
...
-    - nvidia/label/cuda-11.8.0
-    - nvidia
-    - nvidia/label/dev # For pre-releases of MRC. Should still default to full releases if available
-    - conda-forge
+  - rapidsai
+  - nvidia/label/cuda-11.8.0
+  - nvidia
+  - nvidia/label/dev # For pre-releases of MRC. Should still default to full releases if available
+  - conda-forge
+  - numba
```
[the remark this answers is not in the document]
Comment (mdemoret-nv) [reply docker/conda/environments/cuda11.8_dev.yml]: When installing locally, I have no problems ommitting this channel and I see no packages from the `numba` channel installed when I do include it. Do you see something different? I would prefer not to include this new channel since they provide many conflicting packages since as pip, scipy, setuptools, sphinx and the cudatoolkit. Its just asking for issues down the line when we suddenly get a different package than expecting.

File: morpheus/utils/column_info.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (mdemoret-nv) [reply morpheus/utils/column_info.py]: I think we should discuss this before merging. I think having both `get_pandas_dtype()` and `dtype` on the object is very confusing and can easily lead to errors if not used correctly. And `get_pandas_dtype` is only used in 6 places

**Review by mdemoret-nv — APPROVED** [review]
Looks good now. Thanks for all the hard work on this @drobison00

File: docker/conda/environments/cuda11.8_dev.yml (hunk 26 above)
[the remark this answers is not in the document]
Comment (mdemoret-nv) [reply docker/conda/environments/cuda11.8_dev.yml]: Removed the numba channel

File: morpheus/utils/column_info.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (mdemoret-nv) [reply morpheus/utils/column_info.py]: Ended up just using `__set_attr__` since pylint didnt like the descriptors in a dataclass
